Add functional Mamba prefill model

esmalTT opened the pull request.

kpaigwar committed:
#9434: brought up new depthwise conv1d op for mamba

esmalTT committed:
#0: Add torch prefill model

esmalTT committed:
#0: Add functional prefix scan with fallback ops

kpaigwar committed:
#0: extended functional prefill implementation to Mamba Block

esmalTT commented on models/demos/mamba/tests/test_mamba_block.py:
```diff
@@ -82,3 +86,72 @@ def test_mamba_block_inference(
         assert does_pass, f"PCC value is lower than {pcc}"
+
+
+class PytorchMambaBlockPrefill(torch.nn.Module):
```
We should be consistent with the naming in the other test. Maybe `PytorchPrefillMambaBlock` or change the SSM test to `PytorchMambaSSMPrefill`.

kpaigwar replied:
I like the `PytorchPrefillMambaBlock`.

esmalTT commented on models/demos/mamba/tests/test_mamba_ssm.py:
```diff
@@ -51,6 +56,8 @@ def test_mamba_ssm_inference(
     reference_model = MambaDecode.from_pretrained(model_version)
     reference_model.args.batch_size = batch
+    reference_model.args.seq_len = 1
+    reference_model.args.mode = ModelMode.DECODE
```
Adding a 'mode' to the reference model is kinda confusing since it doesn't support both. Can we make the model mode a member of MambaTT instead? We could pass it during initialization and default it to decode.

esmalTT replied:
Maybe combine prefill_model.py and decode_model.py then?

esmalTT commented on models/demos/mamba/tt/mamba_one_step_ssm.py:
```diff
@@ -24,6 +26,8 @@ def __init__(self, args: ModelArgs, device, configs, load_fn: Callable):
         self.configs = configs
         self.n = 32
         self.rank = self.args.dt_rank
+        self.seq_len = args.seq_len
```
Do we need to store this? I think we could just infer sequence length from the input in `forward`, that way we can reducing the amount of state we are keeping.

If we only support certain sequence lengths we can just assert it in `forward`

esmalTT replied:
I guess maybe we need to store it because of the `return x.repeat(self.batch_size * self.seq_len, 1)` on the A parameter...

Should we just fix the seq_len member variable to 32  for now and don't actually pass it as a arg?

kpaigwar replied:
I think we should take care of seq_len here itself as we most likely have to do prefill for seq_len>32 in the functional implementation. Getting error on tt conv for seq_len=32

esmalTT commented on models/demos/mamba/reference/prefix_scan.py:
```diff
@@ -0,0 +1,28 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
I think this needs to be:

```suggestion
# SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
```

kpaigwar committed:
#0: forced asymmetric padding in conv

esmalTT committed:
#9553: Add prefix scan op for Mamba prefill

kpaigwar committed:
#9585: fixed bug in grid calculation when num_cores_nhw<= device_grid_size[0]

kpaigwar committed:
#0: integrated depthwise conv in mamba model

kpaigwar committed:
#0: full model prefill implementation

kpaigwar committed:
#0: added full model test for prefill

esmalTT committed:
#0: Support longer sequence legnths in ssm_prefix_scan

This change adds support for L > 32 in `ssm_eltwise_mul`. Logically this
op can now handle any value of L but values of L > 128 will run out of
L1 in `bfloat8` format.

This change also fixes #9831.

kpaigwar committed:
#0: integrated ttnn prefix scan

kpaigwar committed:
#0: increase num users in ssm_eltwise_mul

kpaigwar committed:
#0: increase num users in ssm_eltwise_mul

kpaigwar committed:
#0: ssm_eltwise_mul working for seq_len>32

kpaigwar committed:
#0: integrated ttnn prefix scan and working with all layers

kpaigwar committed:
#0: ssm_1d_reduce works for the batch>32

esmalTT committed:
#0: Improve softplus kernel accuracy

kpaigwar committed:
#0: prefill model working with seq_len=128
